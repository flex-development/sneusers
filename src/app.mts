/**
 * @file app
 * @module sneusers/app
 */

import useSwagger from '#hooks/use-swagger.hook'
import AppModule from '#modules/app.module'
import { ConsoleLogger } from '@nestjs/common'
import { NestFactory } from '@nestjs/core'
import {
  FastifyAdapter,
  type NestFastifyApplication
} from '@nestjs/platform-fastify'
import type { FastifyReply } from 'fastify'

export default app

/**
 * Create a NestJS application.
 *
 * @see {@linkcode NestFastifyApplication}
 * @see https://fastify.dev
 *
 * @todo versioning
 *
 * @async
 *
 * @this {void}
 *
 * @return {Promise<NestFastifyApplication>}
 *  NestJS x Fastify application
 */
async function app(this: void): Promise<NestFastifyApplication> {
  /**
   * The NestJS application.
   *
   * @const {NestFastifyApplication} app
   */
  const app: NestFastifyApplication = await NestFactory.create(
    AppModule,
    new FastifyAdapter(),
    {
      abortOnError: false,
      logger: new ConsoleLogger({ logLevels: ['verbose'], prefix: 'nest' })
    }
  )

  // decorate the fastify response object.
  app
    .getHttpAdapter()
    .getInstance()
    .decorateReply('end', end) // required for @nestjs/passport
    .decorateReply('setHeader', setHeader) // required for @nestjs/passport

  // configure openapi documentation endpoints.
  useSwagger(app)

  return app
}

/* v8 ignore start */

/**
 * Signal that no more data will be written.
 *
 * @this {FastifyReply}
 *
 * @return {undefined}
 */
function end(this: FastifyReply): undefined {
  return void this.raw.end()
}

/**
 * Set a single header value.
 *
 * If the header already exists in the to-be-sent headers, its value will be
 * replaced.
 * Use an array of strings to send multiple headers with the same name.
 *
 * @this {FastifyReply}
 *
 * @param {string} key
 *  The header name
 * @param {ReadonlyArray<string> | number | string} value
 *  The header value
 * @return {undefined}
 */
function setHeader(
  this: FastifyReply,
  key: string,
  value: number | string | readonly string[]
): undefined {
  return void this.raw.setHeader(key, value)
}

/* v8 ignore end */
