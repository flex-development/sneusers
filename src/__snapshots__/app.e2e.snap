// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e:app > GET / > should respond with api documentation (json) 1`] = `
{
  "openapi": "3.0.0",
  "info": {
    "title": "@flex-development/sneusers",
    "description": "Demo authentication and authorization system",
    "version": "0.0.0",
    "contact": {}
  },
  "servers": [],
  "tags": [],
  "paths": {
    "/accounts": {
      "post": {
        "operationId": "accounts-create",
        "parameters": [],
        "requestBody": {
          "required": true,
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/CreateAccountCommand"
              }
            }
          }
        },
        "responses": {
          "201": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/AccountCreatedPayload"
                }
              }
            }
          },
          "400": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/ValidationException"
                }
              }
            }
          },
          "409": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/EmailConflictException"
                }
              }
            }
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InternalServerException"
                }
              }
            }
          }
        },
        "tags": [
          "accounts"
        ]
      }
    },
    "/accounts/{uid}": {
      "delete": {
        "operationId": "accounts-delete",
        "parameters": [
          {
            "name": "uid",
            "required": true,
            "in": "path",
            "description": "id of account to delete",
            "schema": {
              "type": "string"
            }
          }
        ],
        "responses": {
          "204": {},
          "401": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InvalidCredentialException"
                }
              }
            }
          },
          "403": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/AccessDeniedException"
                }
              }
            }
          },
          "404": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/MissingAccountException"
                }
              }
            }
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InternalServerException"
                }
              }
            }
          }
        },
        "security": [
          {
            "jwt": []
          }
        ],
        "tags": [
          "accounts"
        ]
      }
    },
    "/accounts/whoami": {
      "get": {
        "operationId": "accounts-whoami",
        "parameters": [
          {
            "name": "authorization",
            "in": "header",
            "description": "bearer auth token",
            "required": false,
            "schema": {
              "type": "string"
            }
          }
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/WhoamiPayload"
                }
              }
            }
          },
          "401": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/WhoamiPayload"
                }
              }
            }
          },
          "500": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/InternalServerException"
                }
              }
            }
          }
        },
        "tags": [
          "accounts"
        ]
      }
    }
  },
  "components": {
    "securitySchemes": {
      "bearer": {
        "scheme": "bearer",
        "bearerFormat": "JWT",
        "name": "jwt",
        "type": "http"
      }
    },
    "schemas": {
      "AccessDeniedException": {
        "type": "object",
        "properties": {
          "code": {
            "type": "number",
            "description": "http response status code",
            "enum": [
              403
            ]
          },
          "id": {
            "type": "string",
            "description": "unique id representing the exception",
            "enum": [
              "accounts/access-denied"
            ]
          },
          "message": {
            "type": "string",
            "description": "human-readable description of the exception"
          },
          "reason": {
            "type": "null"
          }
        },
        "required": [
          "code",
          "id",
          "message",
          "reason"
        ]
      },
      "AccountCreatedPayload": {
        "type": "object",
        "properties": {
          "access_token": {
            "type": "string",
            "description": "token for authenticating requests"
          },
          "refresh_token": {
            "type": "string",
            "description": "token for renewing access tokens"
          },
          "uid": {
            "type": "string",
            "description": "unique account id"
          }
        },
        "required": [
          "access_token",
          "refresh_token",
          "uid"
        ]
      },
      "CreateAccountCommand": {
        "type": "object",
        "properties": {
          "email": {
            "type": "string",
            "description": "email address of new user",
            "format": "email"
          },
          "password": {
            "type": "string",
            "description": "password for account",
            "format": "password",
            "minLength": 6
          },
          "type": {
            "type": "string",
            "description": "type of account to create",
            "enum": [
              "developer",
              "user"
            ]
          }
        },
        "required": [
          "email",
          "password",
          "type"
        ]
      },
      "EmailConflict": {
        "type": "object",
        "properties": {
          "email": {
            "type": "string",
            "description": "the conflicting email address"
          }
        },
        "required": [
          "email"
        ]
      },
      "EmailConflictException": {
        "type": "object",
        "properties": {
          "code": {
            "type": "number",
            "description": "http response status code",
            "enum": [
              409
            ]
          },
          "id": {
            "type": "string",
            "description": "unique id representing the exception",
            "enum": [
              "accounts/email-conflict"
            ]
          },
          "message": {
            "type": "string",
            "description": "human-readable description of the exception"
          },
          "reason": {
            "$ref": "#/components/schemas/EmailConflict"
          }
        },
        "required": [
          "code",
          "id",
          "message",
          "reason"
        ]
      },
      "InternalServerException": {
        "type": "object",
        "properties": {
          "code": {
            "type": "number",
            "description": "http response status code",
            "enum": [
              500
            ]
          },
          "id": {
            "type": "string",
            "description": "unique id representing the exception",
            "enum": [
              "sneusers/internal-error"
            ]
          },
          "message": {
            "type": "string",
            "description": "human-readable description of the exception"
          },
          "reason": {
            "type": "null"
          }
        },
        "required": [
          "code",
          "id",
          "message",
          "reason"
        ]
      },
      "InvalidCredentialException": {
        "type": "object",
        "properties": {
          "code": {
            "type": "number",
            "description": "http response status code",
            "enum": [
              401
            ]
          },
          "id": {
            "type": "string",
            "description": "unique id representing the exception",
            "enum": [
              "accounts/invalid-credential"
            ]
          },
          "message": {
            "type": "string",
            "description": "human-readable description of the exception"
          },
          "reason": {
            "type": "null"
          }
        },
        "required": [
          "code",
          "id",
          "message",
          "reason"
        ]
      },
      "MissingAccount": {
        "type": "object",
        "properties": {
          "uid": {
            "type": "string",
            "description": "id of missing account"
          }
        },
        "required": [
          "uid"
        ]
      },
      "MissingAccountException": {
        "type": "object",
        "properties": {
          "code": {
            "type": "number",
            "description": "http response status code",
            "enum": [
              404
            ]
          },
          "id": {
            "type": "string",
            "description": "unique id representing the exception",
            "enum": [
              "accounts/not-found"
            ]
          },
          "message": {
            "type": "string",
            "description": "human-readable description of the exception"
          },
          "reason": {
            "$ref": "#/components/schemas/MissingAccount"
          }
        },
        "required": [
          "code",
          "id",
          "message",
          "reason"
        ]
      },
      "ValidationException": {
        "type": "object",
        "properties": {
          "code": {
            "type": "number",
            "description": "http response status code",
            "enum": [
              400
            ]
          },
          "id": {
            "type": "string",
            "description": "unique id representing the exception",
            "enum": [
              "sneusers/validation-failure"
            ]
          },
          "message": {
            "type": "string",
            "description": "human-readable description of the exception"
          },
          "reason": {
            "$ref": "#/components/schemas/ValidationFailure"
          }
        },
        "required": [
          "code",
          "id",
          "message",
          "reason"
        ]
      },
      "ValidationFailure": {
        "type": "object",
        "properties": {
          "constraints": {
            "type": "object",
            "additionalProperties": {
              "type": "string"
            },
            "description": "violated validation constraints"
          },
          "property": {
            "type": "string",
            "description": "the name of the property that caused the validation failure"
          },
          "value": {
            "description": "the property value that caused the validation failure",
            "oneOf": [
              {
                "type": "array"
              },
              {
                "type": "boolean"
              },
              {
                "type": "null"
              },
              {
                "type": "number"
              },
              {
                "type": "object"
              },
              {
                "type": "string"
              }
            ]
          }
        },
        "required": [
          "constraints",
          "property"
        ]
      },
      "WhoamiPayload": {
        "type": "object",
        "properties": {
          "uid": {
            "description": "unique account id",
            "oneOf": [
              {
                "type": "string"
              },
              {
                "type": "null"
              }
            ]
          }
        },
        "required": [
          "uid"
        ]
      }
    }
  }
}
`;
